### PR TITLE
[3030] Add health check endpoint to test DB,Redis & Sidekiq

### PR DESCRIPTION
**Pull request by aliuk2012**
### Context
We need an endpoint that we can use to test that the API's DB, Redis Instance and Sidekiq is all working. We will use this to alert us when there are issues with the platform.

As a member of the team
I would like to know when the app is having an outage
So that I can respond as soon as possible

Add healthcheck endpoints to our services that report the status of any dependencies the app has, i.e. DB connection, Redis connection, API connection (using /ping), etc.

### Success

* We have an endpoint that checks that the API can connect to:
  * DB
  * Redis
  * sidekiq
* The status is emitted as JSON.

### Changes proposed in this pull request
Returns the following JSON:

```json
{
  "checks": {
    "database": true,
    "redis": true,
    "sidekiq": true,
    "sidekiq_queue": true
  }
}
```
### Guidance to review
run `bundle exec rails s` and then in the browser go to `localhost:3001/healthcheck` to see the results. Try stopping your db,redis and sidekiq and test the results.
### Checklist

- [x] Make sure all information from the Trello card is in here
- [x] Attach to Trello card
- [x] Rebased master
- [x] Cleaned commit history
- [x] Tested by running locally

### CODE REVIEW

**Comment by aliuk2012** [comment]
[Add /healthcheck endpoint to ttapi](https://trello.com/c/XNOMhtNQ/3030-add-healthcheck-endpoint-to-ttapi)

**Comment by aliuk2012** [comment]
[Add /healthcheck endpoint to ttapi](https://trello.com/c/XNOMhtNQ/3030-add-healthcheck-endpoint-to-ttapi)

**Comment by aliuk2012** [comment]
[Add /healthcheck endpoint to ttapi](https://trello.com/c/XNOMhtNQ/3030-add-healthcheck-endpoint-to-ttapi)

**Comment by aliuk2012** [comment]
[Add /healthcheck endpoint to ttapi](https://trello.com/c/XNOMhtNQ/3030-add-healthcheck-endpoint-to-ttapi)

**Comment by aliuk2012** [comment]
[Add /healthcheck endpoint to ttapi](https://trello.com/c/XNOMhtNQ/3030-add-healthcheck-endpoint-to-ttapi)

**Comment by TobyRet** [comment]
@aliuk2012 - are there plans to set up a cron job to ping the health check endpoint? What will we be using to monitor success / failure?

**Comment by aliuk2012** [comment]
I'm not 100% sure, I'll leave @misaka to answer those very valid questions.

**Comment by timabell** [comment]
I think I like `/ping` as an endpoint, it's easy to remember

**Comment by timabell** [comment]
I'd squash all those commits into one.

**Comment by aliuk2012** [comment]
I was going to repurpose the existing `/ping` endpoint but decided against it for two reasons
- I wasn't sure where we were using it and making changes might break things
- `/ping` could be a very cheap way just to check a connection can be made without querying activerecord/sidekiq and redis.

**Comment by aliuk2012** [comment]
I'll squash them once this PR is ready to be merged

**Comment by timabell** [comment]
Works for me :heart: 
Good job!

```
$ curl localhost:3001/healthcheck
{"checks":{"database":true,"redis":false,"sidekiq":false,"sidekiq_queue":false}}%                                                                                                
```

File: app/controllers/heartbeat_controller.rb [hunk 1]
```diff
@@ -0,0 +1,51 @@
+require "sidekiq/api"
+
+class HeartbeatController < ActionController::API
+  def ping
```
Comment (timabell) [inline app/controllers/heartbeat_controller.rb]: do we need this as well as `healthcheck`? seems redundant

**Comment by timabell** [comment]
I think we should kill the old ping-pong unless we know for sure it's needed. Clean code and all that. We can still tell the difference between dependency failure (`502`) and complete service failure (`500` or connection failure) with the new endpoint.

**Comment by timabell** [comment]
* front-door spams the hell out of `/ping` so wouldn't want to make that expensive
* swap-slot uses ping too, would be good for that to switch to the advanced healthcheck in this PR

:ship: it

File: app/controllers/heartbeat_controller.rb (hunk 1 above)
Comment (misaka) [reply app/controllers/heartbeat_controller.rb]: I say we keep it for stuff like Frontdoor that pings several times a second.

File: app/controllers/heartbeat_controller.rb [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+  def healthcheck
+    checks = {
+      database: database_alive?,
+      redis: redis_alive?,
+      sidekiq: sidekiq_alive?,
+      sidekiq_queue: sidekiq_queue_healthy?,
+    }
+
+    status = :bad_gateway unless checks.values.all?
```
Comment (misaka) [inline app/controllers/heartbeat_controller.rb]: This could be clearer that we'll end up with a success code.

**Pull request merged**